File: Distribution-Grid-EV-CA/R_Yanning/07_05_grid_EV.R
library(data.table)

load.PGE <- fread('data/grid data/clean/clean load PGE_full hour.csv')
load.SCE <- fread('data/grid data/clean/clean load SCE_fixed.csv')
load.SDGE <- fread('data/grid data/clean/clean load SDGE_2022.csv')


cap.PGE <- fread('data/grid data/clean/clean ICA PGE_all.csv')
cap.SCE <- fread('data/grid data/clean/clean GNA SCE.csv')
cap.SDGE <- fread('data/grid data/clean/clean ICA SDGE_all.csv')

EV <- readRDS('data/result/charging profile by feeder/profile_feeder_cum_bins_IOU_sample42_draw12_split6699_sample36.rds')
head(EV)
summary(EV)

setnames(cap.PGE,"NetworkId","FeederID")
setnames(cap.PGE,"Month","month")
setnames(cap.PGE,"Hour","hour")
setnames(cap.PGE,"capacity_kW","headroom")
PGE <- merge(x=load.PGE[,c("FeederID","load_high",'month',"hour")],y=cap.PGE,
             by=c("FeederID","month","hour"))
setnames(PGE,"load_high","baseload")
PGE[,capacity:=headroom+baseload]
PGE[,FeederID:=as.character(FeederID)]
unique(nchar(PGE$FeederID))
PGE[nchar(FeederID)==8,FeederID:=paste0('0',FeederID)]
PGE[,utility:='PGE']
summary(PGE)
PGE[capacity<0]

SCE <- merge(x=load.SCE,y=cap.SCE[,c('FeederID',"capacity_kW")],
             by='FeederID')
setnames(SCE,"load_high","baseload")
setnames(SCE,"capacity_kW","capacity")
SCE[,headroom:=capacity-baseload]
SCE[,utility:='SCE']
summary(SCE)
SCE[headroom<0]

unique(nchar(load.SDGE$FeederID))
cap.SDGE[,FeederID:=as.character(FeederID)]
SDGE <- merge(x=load.SDGE,y=cap.SDGE[,c("FeederID",'month','hour','capacity')],
              by=c("FeederID","month","hour"))
setnames(SDGE,'capacity','headroom')
setnames(SDGE,'load','baseload')
SDGE[,capacity:=headroom+baseload]
SDGE[,utility:='SDGE']
summary(SDGE)
SDGE[capacity<0]
SDGE[capacity>16000]

grid <- rbind(PGE,SCE,SDGE)

EV.sum <- EV[,.(EVload=sum(load)),by=.(FeederID,year,hour)]
EV.sum <- EV.sum[rep(1:nrow(EV.sum),each=12),]
EV.sum[,month:=rep(1:12,times=nrow(EV.sum)/12)]
head(EV.sum)

EV.grid <- merge(x=EV.sum,y=grid,
                 by=c('FeederID','month','hour'))
head(EV.grid)
length(unique(EV.grid$FeederID)) # 5582 valid feeders in total
EV.grid[,overload:=EVload-headroom]
nrow(EV.grid[year==2022&overload>0]) # 34557 hours
nrow(EV.grid[year==2045&overload>0]) # 496593 hours
length(unique(EV.grid[year==2022&headroom<=0]$FeederID)) # 654/5582 (data issues)
length(unique(EV.grid[year==2022&overload>0]$FeederID)) # 795/5582
length(unique(EV.grid[year==2045&overload>0]$FeederID)) # 4452/5582

saveRDS(EV.grid,'data/result/overload/overload_sample42_draw12_split6699_sample36.rds')

File: Distribution-Grid-EV-CA/07_05_grid_EV.py
"""
07_05_grid_EV.py — Python port of ``07_05_grid_EV.R``.

Combines utility baseline load & capacity (PGE, SCE, SDGE) with the cumulative
EV charging load per feeder to produce an overload table.
"""

from __future__ import annotations

import os

import numpy as np
import pandas as pd

from common import read_rds_like, save_rds_like

LOAD_PGE = "data/grid data/clean/clean load PGE_full hour.csv"
LOAD_SCE = "data/grid data/clean/clean load SCE_fixed.csv"
LOAD_SDGE = "data/grid data/clean/clean load SDGE_2022.csv"

CAP_PGE = "data/grid data/clean/clean ICA PGE_all.csv"
CAP_SCE = "data/grid data/clean/clean GNA SCE.csv"
CAP_SDGE = "data/grid data/clean/clean ICA SDGE_all.csv"

EV_PKL = (
    "data/result/charging profile by feeder/"
    "profile_feeder_cum_bins_IOU_sample42_draw12_split6699_sample36.pkl"
)
OUT_PKL = (
    "data/result/overload/overload_sample42_draw12_split6699_sample36.pkl"
)


def _prep_pge() -> pd.DataFrame:
    load = pd.read_csv(LOAD_PGE)[["FeederID", "load_high", "month", "hour"]]
    cap = pd.read_csv(CAP_PGE).rename(
        columns={"NetworkId": "FeederID", "Month": "month", "Hour": "hour", "capacity_kW": "headroom"}
    )
    pge = pd.merge(load, cap, on=["FeederID", "month", "hour"]).rename(columns={"load_high": "baseload"})
    pge["capacity"] = pge["headroom"] + pge["baseload"]
    pge["FeederID"] = pge["FeederID"].astype(str)
    pge.loc[pge["FeederID"].str.len() == 8, "FeederID"] = "0" + pge.loc[pge["FeederID"].str.len() == 8, "FeederID"]
    pge["utility"] = "PGE"
    return pge


def _prep_sce() -> pd.DataFrame:
    load = pd.read_csv(LOAD_SCE)
    cap = pd.read_csv(CAP_SCE)[["FeederID", "capacity_kW"]]
    sce = pd.merge(load, cap, on="FeederID").rename(
        columns={"load_high": "baseload", "capacity_kW": "capacity"}
    )
    sce["headroom"] = sce["capacity"] - sce["baseload"]
    sce["utility"] = "SCE"
    return sce


def _prep_sdge() -> pd.DataFrame:
    load = pd.read_csv(LOAD_SDGE)
    cap = pd.read_csv(CAP_SDGE)[["FeederID", "month", "hour", "capacity"]]
    cap["FeederID"] = cap["FeederID"].astype(str)
    sdge = pd.merge(load, cap, on=["FeederID", "month", "hour"]).rename(
        columns={"capacity": "headroom", "load": "baseload"}
    )
    sdge["capacity"] = sdge["headroom"] + sdge["baseload"]
    sdge["utility"] = "SDGE"
    return sdge


def main() -> None:
    os.makedirs(os.path.dirname(OUT_PKL), exist_ok=True)
    ev = read_rds_like(EV_PKL)

    pge = _prep_pge()
    sce = _prep_sce()
    sdge = _prep_sdge()
    grid = pd.concat([pge, sce, sdge], ignore_index=True)

    ev_sum = (
        ev.groupby(["FeederID", "year", "hour"], as_index=False)["load"]
        .sum()
        .rename(columns={"load": "EVload"})
    )
    # replicate per month 1..12
    ev_sum = ev_sum.loc[ev_sum.index.repeat(12)].reset_index(drop=True)
    n_rows = len(ev_sum) // 12
    ev_sum["month"] = np.tile(np.arange(1, 13), n_rows)

    ev_grid = pd.merge(ev_sum, grid, on=["FeederID", "month", "hour"])
    ev_grid["overload"] = ev_grid["EVload"] - ev_grid["headroom"]
    save_rds_like(ev_grid, OUT_PKL)

    print("valid feeders:", ev_grid["FeederID"].nunique())


if __name__ == "__main__":
    main()
